tasks/src/tasks/fetch.rs

```diff
@@ -0,0 +1,116 @@
+    fetch::{MacroOrMessageDataItemNames, MessageDataItem},
+    response::{Data, StatusBody, StatusKind},
+    sequence::SequenceSet,
+};
+use tracing::warn;
+
+use super::TaskError;
+use crate::Task;
+
+/// Fetch message data items matching the given sequence set and the
+/// given item names (or macro).
+#[derive(Clone, Debug)]
+pub struct FetchTask {
+    sequence_set: SequenceSet,
+    macro_or_item_names: MacroOrMessageDataItemNames<'static>,
+    uid: bool,
+    output: HashMap<NonZeroU32, Vec1<MessageDataItem<'static>>>,
+}
+
+impl Task for FetchTask {
+    type Output = Result<HashMap<NonZeroU32, Vec1<MessageDataItem<'static>>>, TaskError>;
+
+    fn command_body(&self) -> CommandBody<'static> {
+        CommandBody::Fetch {
+            sequence_set: self.sequence_set.clone(),
+            macro_or_item_names: self.macro_or_item_names.clone(),
+            uid: self.uid,
+        }
+    }
+
+    fn process_data(&mut self, data: Data<'static>) -> Option<Data<'static>> {
+        if let Data::Fetch { items, seq } = data {
+            if let Some(items) = self.output.insert(seq, items) {
+                warn!(?items, "received duplicate items for message {seq}");
+            }
+
+            None
+        } else {
+            Some(data)
+        }
+    }
+
+    fn process_tagged(self, status_body: StatusBody<'static>) -> Self::Output {
+        match status_body.kind {
+            StatusKind::Ok => Ok(self.output),
+            StatusKind::No => Err(TaskError::UnexpectedNoResponse(status_body)),
+            StatusKind::Bad => Err(TaskError::UnexpectedBadResponse(status_body)),
+        }
+    }
+}
+
+impl FetchTask {
+    pub fn new(sequence_set: SequenceSet, items: MacroOrMessageDataItemNames<'static>) -> Self {
+        Self {
+            sequence_set,
+            macro_or_item_names: items,
+            uid: true,
+            output: Default::default(),
+        }
+    }
+
+    pub fn set_uid(&mut self, uid: bool) {
+        self.uid = uid;
+    }
+
+    pub fn with_uid(mut self, uid: bool) -> Self {
+        self.set_uid(uid);
+        self
+    }
+
+    pub fn first(self) -> FetchFirstTask {
+        FetchFirstTask(self)
+    }
+}
+
+/// Same as [`FetchTask`], except that it only collects message data
+/// items for the first message matching the given sequence set.
+#[derive(Clone, Debug)]
+pub struct FetchFirstTask(FetchTask);
+
+impl FetchFirstTask {
+    pub fn new(fetch: FetchTask) -> Self {
+        Self(fetch)
+    }
+}
+
+impl Task for FetchFirstTask {
+    type Output = Result<Option<Vec1<MessageDataItem<'static>>>, TaskError>;
+
+    fn command_body(&self) -> CommandBody<'static> {
+        self.0.command_body()
+    }
+
+    fn process_data(&mut self, data: Data<'static>) -> Option<Data<'static>> {
+        match data {
+            Data::Fetch { items, seq } if self.0.output.is_empty() => {
+                self.0.output.insert(seq, items);
+                None
+            }
+            data => Some(data),
```

Comment:
Ah, I see. If I understand correctly, the remaining `Fetch`s will become `unsolicited` data now? I think this could be confusing...

What I also find not ideal is that this tasks seems to fetch at most one message but really it fetches more.

Maybe it's better to do the filtering at a later stage for now? This would make obvious what happens (and reduce duplicated code for now)? Alternatively, we should maybe think about avoiding the overfetches? Sadly, I don't know how to do this in a single task.

What do you think?

Reply:
I pushed an alternative, what do you think about? If it does not fit, I agree to filter before and after, around the `FetchTask`, it also makes sense. I just found convenient the fact to propose a `FetchFirstTask` (somehow I had this need at higher level).

Reply:
I'm a bit tired, so I hope I understand right. The `FetchFirstTask` now has a single seq/uid, right? Is it worth to duplicate the task code for this case? Because: In the end, it's similar to calling `FetchTask` with `seq.iter().first()` instead `seq`? Do I understand right?

What could be a useful task is something like "Fetch the first message from 1:*" (what you had before) BUT with some smarter commands, e.g., first try to find out which message is the first (w/o fetching the others) and then only fetching the first one. But this is currently not possible to do in a single `Task` because it's bound to a single `CommandBody`.

Reply:
> Is it the same as calling `FetchTask` with `seq.iter().first()` for a specific `seq`?

Technically yes, but I don't like the fact that running `FetchTask` with one unique specific sequence number or uid (not a range) outputs a hash map. By running `FetchFirstTask`, you can only define 1 seq/uid, therefore you get 1 fetch. In this particular case, I value more the public API over code duplication.

Maybe the naming induces us in error. `FetchFirstTask` implies that you take *the first one from a list*. What about:

- `FetchTask` becomes one of `FetchManyTask`, `RangeFetchTask`
- `FetchFirstTask` becomes one of `FetchTask`, `FetchSingleTask`, `FetchOneTask`

Fetching the first task of a range can be done at the higher level client, as you said.

> But this is currently not possible to do in a single `Task` because `Task` is bound to a single `CommandBody`.

I guess it is sth to do in the higher level client as well?

Sth like could be interesting also is a `FetchAllTask`, which is just a `FetchManyTask` with `..` sequence range.